### DISCUSSION TITLE
RFC: Embedding application security and sandboxing blocks

### DISCUSSION THREAD

**Discussion by vilkinsons** [blockprotocol/blockprotocol#181] [RFC]
## The need for sandboxing

- Preserving data privacy and confidentiality in the face of malicious or badly-built blocks
- Preserving data integrity
- Preventing side-effects (e.g. blocks leaking styles or interfering with Embedding Application functionality)

The Block Protocol spec [specifically calls out](https://blockprotocol.org/spec/embedding-applications#security) the security concern inherent in running arbitrary code from untrusted third-parties. It mentions the iFrame-based approach to sandboxing employed by our example embedding application, HASH, whose source code is public. We do not (yet) however provide inline docs and examples re: how embedding applications can harden themselves against malicious actors and protect their users from ne'er-do-wellers. We intend to. This experimental RFC captures the matter as an outstanding area for expansion, as well as our thoughts.

We've additionally been exploring alternative approaches to the iFrame/postMessage paradigm, in particular how WebAssembly might be used. We'd welcome contributions from community members and interested parties re: either, as well as further/additional, emerging or novel approaches not already considered.

**HASH's BP working group is of the opinion that sandboxing should be recommended by default, but not enforced.**

## Underpinning Assumptions

- it will be generally undesirable to remove the sandbox around blocks (whether due to dependency management complexities, security risks or side effects)
- there will be instances where embedding applications wish to unsandbox trusted sets of blocks with common sets of dependencies (e.g. a 'HASH library' of core blocks which provides commonly relied upon blocks).
- furthermore, the use of blocks by users will largely follow a power-law distribution in which the most-used handful of blocks are _far more used_ than any other type of block (perhaps comprising 90%+ of total embeds). The vast majority of the blocks on any given page in a CMS or knowledge management application context will be generic text, image, or data-rendering blocks (e.g. table/calendar/columnar board). These may be provided from one or more unsandboxed core libraries which can be trusted to act responsibly (e.g. not cause side effects, bleed styles, etc.) and allowed to act synchronously.
- although comprising a minority of embeds, the remaining 'long tail' of blocks provides great utility to users reliant upon them when they are called upon, and will be invaluable. Great utility is unlocked by the minority of blocks of the page which provide specialist functionality. For example:
  - data blocks designed to render specific types of information in a tailored view, for example a [person contact card](https://blockprotocol.org/@hash/blocks/person)
  - utility-providing blocks such as a [calendar appointment booking block](https://github.com/blockprotocol/blockprotocol/issues/201)).
- it is impractical for embedding applications to build and maintain thousands of long-tail blocks used by folks in specific use-cases of their products
- the Block Protocol provides applications with an ability to allow users to tailor them for their own needs, and invent their own workflows

**Comment by tomsdev** [comment]
If Blocks components were written in a programing language that doesn't allow global mutations, side effects, or to access the network, file system, etc. Then I don't think you would need to sandbox/iframe these blocks components in the embeding application.

On the frontend, Elm is one of the programing language I know would work here. Blocks components could be written in Elm and then embedded in a Typescript/JavaScript application and that would be safe.

On a side note, on the backend, both Deno and node-safe managed to make running JavaScript/Typescript safer.

**Reply by acruise** [reply]
> If Blocks components were written in a programing language that doesn't allow global mutations, side effects, or to access the network, file system, etc.

I've been looking at non-turing-complete languages for these kinds of semi-trusted-user things for a few years. I know of two examples that seem worth trying:
* [CEL](https://github.com/google/cel-go) -- not just golang, there are a few more implementations. It has its own syntax, uses Protobuf for the untyped and typed AST, and is really non-Turing-complete, with a macro system to support bounded iteration
* [Starlark](https://github.com/bazelbuild/starlark) -- not just JVM, there are a few more implementations. Python-like syntax, and actually Turing-complete, but uses a step/cycle counter so the interpreter can abort if something appears to be spinning forever

Both provide no built-in support for IO or externally visible side effects, but both also allow values and functions, written in the host language, to be injected into their namespace so they can do something useful.

**Reply by DougAnderson444** [reply]
Interesting to consider non-JS languages. But how do you get that slick user interactivity in the browser if you aren't using JavaScript?

**Comment by jtewright** [comment]
I work at HASH/Block Protocol and wanted to share a quick update on this topic.

We announced [version 0.2 of the Block Protocol](https://hash.dev/blog/block-protocol-v02) yesterday. It introduces a new architecture which will help make the protocol more extensible.

As part of the work, we spent a lot of time thinking about sandboxing. The thinking is still in progress, but within the new **RFCs & Roadmap** section of the new spec, there's an explanation of where we are: https://blockprotocol.org/docs/spec/rfcs_and_roadmap#core

We'd love any thoughts and feedback you have on this!

Thank you
